File: BACKSPRING/backspring/src/main/java/com/iaspring/backspring/service/IaJugeService.java
package com.iaspring.backspring.service;

import org.springframework.stereotype.Service;

@Service
public class IaJugeService {

    public boolean evaluerReponse(String reponseAttendue, String synonymes, String reponseJoueur) {
        if (reponseJoueur == null || reponseAttendue == null || reponseJoueur.isEmpty()) return false;

        // 1. On teste d'abord la réponse officielle stricte
        if (testerCorrespondance(reponseAttendue, reponseJoueur)) {
            return true;
        }

        // 2. Si c'est faux, on découpe et on teste la liste des synonymes
        if (synonymes != null && !synonymes.trim().isEmpty()) {
            String[] listeSynonymes = synonymes.split(",");
            for (String mot : listeSynonymes) {
                if (testerCorrespondance(mot.trim(), reponseJoueur)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean testerCorrespondance(String attendu, String test) {
        String cible = nettoyer(attendu);
        String essai = nettoyer(test);

        if (cible.equals(essai)) return true;

        int distance = calculerLevenshtein(cible, essai);

        // 👉 TOLÉRANCE PROPORTIONNELLE (Jusqu'à 5 fautes)
        int tolerance;
        int longueur = cible.length();
        
        if (longueur <= 5) tolerance = 1;       // ex: "Paris" (1 faute max)
        else if (longueur <= 10) tolerance = 2; // ex: "Tennis" (2 fautes max)
        else if (longueur <= 15) tolerance = 3; // ex: "Royaume Uni" (3 fautes max)
        else if (longueur <= 20) tolerance = 4; // ex: "Napoléon Bonaparte" (4 fautes)
        else tolerance = 5;                     // Très longues phrases (5 fautes max)

        return distance <= tolerance;
    }

    private String nettoyer(String input) {
        String clean = input.trim().toLowerCase()
                .replaceAll("[éèêë]", "e")
                .replaceAll("[àâä]", "a")
                .replaceAll("[îï]", "i")
                .replaceAll("[ôö]", "o")
                .replaceAll("[ûüù]", "u")
                .replaceAll("[ç]", "c")
                .replaceAll("[-']", " "); // Remplace les apostrophes par des espaces ("l'eau" devient "l eau")

        // 👉 SUPPRESSION DES ARTICLES : Si le mot commence par le, la, les, un, une... on l'ignore
        clean = clean.replaceAll("^(le |la |les |l |un |une |des )", "");
        
        return clean.trim();
    }

    private int calculerLevenshtein(String a, String b) {
        int[][] dp = new int[a.length() + 1][b.length() + 1];

        for (int i = 0; i <= a.length(); i++) {
            for (int j = 0; j <= b.length(); j++) {
                if (i == 0) {
                    dp[i][j] = j;
                } else if (j == 0) {
                    dp[i][j] = i;
                } else {
                    int cost = (a.charAt(i - 1) == b.charAt(j - 1)) ? 0 : 1;
                    dp[i][j] = Math.min(dp[i - 1][j - 1] + cost, 
                               Math.min(dp[i - 1][j] + 1, 
                                        dp[i][j - 1] + 1));
                }
            }
        }
        return dp[a.length()][b.length()];
    }
}

File: BACKDJANGO/backdjango/game/ia_juge_service.py
import unicodedata
import re

class IaJugeService:

    @staticmethod
    def evaluer_reponse(reponse_attendue: str, synonymes: str, reponse_joueur: str) -> bool:
        if not reponse_joueur or not reponse_attendue or reponse_joueur.strip() == "":
            return False

        # 1. On teste la réponse officielle stricte
        if IaJugeService.tester_correspondance(reponse_attendue, reponse_joueur):
            return True

        # 2. On teste la liste des synonymes
        if synonymes and synonymes.strip():
            liste_synonymes = synonymes.split(',')
            for mot in liste_synonymes:
                if IaJugeService.tester_correspondance(mot.strip(), reponse_joueur):
                    return True

        return False

    @staticmethod
    def tester_correspondance(attendu: str, test: str) -> bool:
        cible = IaJugeService.nettoyer(attendu)
        essai = IaJugeService.nettoyer(test)

        if cible == essai:
            return True

        distance = IaJugeService.calculer_levenshtein(cible, essai)

        # 👉 TOLÉRANCE PROPORTIONNELLE (Jusqu'à 5 fautes)
        longueur = len(cible)
        if longueur <= 5:
            tolerance = 1
        elif longueur <= 10:
            tolerance = 2
        elif longueur <= 15:
            tolerance = 3
        elif longueur <= 20:
            tolerance = 4
        else:
            tolerance = 5

        return distance <= tolerance

    @staticmethod
    def nettoyer(texte: str) -> str:
        # Met en minuscule et retire les accents
        texte = texte.strip().lower()
        texte = ''.join(c for c in unicodedata.normalize('NFD', texte) if unicodedata.category(c) != 'Mn')
        
        # Remplace tirets et apostrophes par des espaces
        texte = re.sub(r"[-']", " ", texte)
        
        # 👉 SUPPRESSION DES ARTICLES au début du mot
        texte = re.sub(r"^(le |la |les |l |un |une |des )", "", texte)
        
        return texte.strip()

    @staticmethod
    def calculer_levenshtein(a: str, b: str) -> int:
        if len(a) < len(b):
            return IaJugeService.calculer_levenshtein(b, a)
        if len(b) == 0:
            return len(a)
        
        previous_row = range(len(b) + 1)
        for i, c1 in enumerate(a):
            current_row = [i + 1]
            for j, c2 in enumerate(b):
                insertions = previous_row[j + 1] + 1
                deletions = current_row[j] + 1
                substitutions = previous_row[j] + (c1 != c2)
                current_row.append(min(insertions, deletions, substitutions))
            previous_row = current_row
        
        return previous_row[-1]